src/erlang_ls_indexer.erl

```diff
@@ -0,0 +1,210 @@
+%% API
+-export([ find_and_index_file/1
+        , find_and_index_file/2
+        , index/1
+        , index_dir/1
+        , start_link/0
+        , index_app/0
+        , index_deps/0
+        , index_otp/0
+        ]).
+
+%% gen_server callbacks
+-export([ init/1
+        , handle_call/3
+        , handle_cast/2
+        , terminate/2
+        ]).
+
+%%==============================================================================
+%% Includes
+%%==============================================================================
+-include("erlang_ls.hrl").
+
+%%==============================================================================
+%% Types
+%%==============================================================================
+-type state() :: #{}.
+
+%%==============================================================================
+%% Macros
+%%==============================================================================
+-define(SERVER, ?MODULE).
+
+%%==============================================================================
+%% Exported functions
+%%==============================================================================
+-spec find_and_index_file(string()) ->
+   {ok, uri()} | {error, any()}.
+find_and_index_file(FileName) ->
+  find_and_index_file(FileName, async).
+
+-spec find_and_index_file(string(), async | sync) ->
+   {ok, uri()} | {error, any()}.
+find_and_index_file(FileName, SyncAsync) ->
+  Paths = lists:append([ erlang_ls_config:get(app_paths)
+                       , erlang_ls_config:get(deps_paths)
+                       , erlang_ls_config:get(otp_paths)
+                       ]),
+  case file:path_open(Paths, list_to_binary(FileName), [read]) of
+    {ok, IoDevice, FullName} ->
+      %% TODO: Avoid opening file twice
+      file:close(IoDevice),
+      try_index_file(FullName, SyncAsync),
+      {ok, erlang_ls_uri:uri(FullName)};
+    {error, Error} ->
+      {error, Error}
+  end.
+
+-spec index(erlang_ls_document:document()) -> ok.
+index(Document) ->
+  Uri    = erlang_ls_document:uri(Document),
+  ok     = erlang_ls_db:store(documents, Uri, Document),
+  Module = erlang_ls_uri:module(Uri),
+  ok = erlang_ls_db:store(modules, Module, Uri),
+  Specs  = erlang_ls_document:points_of_interest(Document, [spec]),
+  [erlang_ls_db:store(signatures, {Module, F, A}, Tree) ||
+    #{data := {{F, A}, Tree}} <- Specs],
+  Kinds = [application, implicit_fun],
+  POIs  = erlang_ls_document:points_of_interest(Document, Kinds),
+  [register_reference(Uri, POI) || POI <- POIs],
+  ok.
+
+-spec index_app() -> any().
+index_app() ->
+  gen_server:cast(?SERVER, {index_app}).
+
+-spec index_deps() -> any().
+index_deps() ->
+  case application:get_env(erlang_ls, index_deps) of
+    {ok, true}  -> gen_server:cast(?SERVER, {index_deps});
+    {ok, false} -> lager:info("Not indexing dependencies")
+  end.
+
+-spec index_otp() -> any().
+index_otp() ->
+  case application:get_env(erlang_ls, index_otp) of
```

Comment:
These were added as application variables, I considered adding them as user configuration, but I thought that the users shouldn't care about this. @robertoaloi WDYT?

Reply:
It is fine as an application env for now.